### PR TITLE
V3 monitoring jobs - add new queue-based jobs

### PR DESCRIPTION
**Pull request by scottbommarito**
Depends on https://github.com/NuGet/NuGet.Services.Metadata/pull/243

Added `Catalog2Monitoring`, `MonitoringProcessor`, and `Monitoring2Monitoring`, the three jobs that make up the V3 monitoring jobs.

`Catalog2Monitoring`:
Runs on the catalog and pushes new commits to a queue.

`MonitoringProcessor`:
Listens to the queue and runs validations on each commit.

`Monitoring2Monitoring`:
Requeues packages that have failed validation.

### CODE REVIEW

**Comment by chenriksson** [comment]
For the dependency on #243, I see the common code.

Why not create this PR against the sb-v3frameworkfixes feature branch instead of dev so that we don't have code duplication in the reviews. This saves time for reviewers and prevents you from getting conflicting advice.

**Comment by scottbommarito** [comment]
@chenriksson Will fix, thanks!

File: src/Ng/Jobs/Catalog2MonitoringJob.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NuGet.Services.Configuration;
+using NuGet.Services.Metadata.Catalog;
+using NuGet.Services.Metadata.Catalog.Monitoring;
+
+namespace Ng.Jobs
+{
+    public class Catalog2MonitoringJob : LoopingNgJob
```
Comment (loic-sharma) [inline src/Ng/Jobs/Catalog2MonitoringJob.cs]: >Catalog2MonitoringJob [](start = 17, length = 21)

For each of these jobs add a comment describing what it is it does.

File: src/Ng/CommandHelpers.cs [hunk 2]
```diff
@@ -79,7 +83,7 @@ private static ISecretInjector GetSecretInjector(IDictionary<string, string> arg
 
         public static RegistrationStorageFactories CreateRegistrationStorageFactories(IDictionary<string, string> arguments, bool verbose)
         {
-            StorageFactory legacyStorageFactory;
+            NuGet.Services.Metadata.Catalog.Persistence.StorageFactory legacyStorageFactory;
```
Comment (loic-sharma) [inline src/Ng/CommandHelpers.cs]: >NuGet.Services.Metadata.Catalog.Persistence.StorageFactory [](start = 12, length = 58)

Please don't put the type's full-name here. If there's a naming conflict, add an alias like `CatalogStorageFactory`.

File: src/Ng/Jobs/MonitoringProcessorJob.cs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+            _client = new CollectorHttpClient(messageHandlerFactory());
+        }
+
+        protected override async Task RunInternal(CancellationToken cancellationToken)
+        {
+            await ParallelExecutionUtility.ProcessInParallel(() => ProcessPackages(cancellationToken));
+        }
+
+        private async Task ProcessPackages(CancellationToken token)
```
Comment (loic-sharma) [inline src/Ng/Jobs/MonitoringProcessorJob.cs]: This method is quite large. Please move pieces out into helpers where possible

Comment (scottbommarito) [reply src/Ng/Jobs/MonitoringProcessorJob.cs]: done

**Comment by shishirx34** [comment]
Could you resolve the conflicts here.

**Comment by scottbommarito** [comment]
@shishirx34 yup, PR got messed up when I merged the one that this depended on because of a squash, woops

File: src/Ng/CommandHelpers.cs [hunk 4]
```diff
@@ -5,17 +5,28 @@
...
 using Microsoft.WindowsAzure.Storage;
 using Microsoft.WindowsAzure.Storage.Auth;
 using NuGet.Services.Configuration;
 using NuGet.Services.KeyVault;
 using NuGet.Services.Metadata.Catalog;
+using NuGet.Services.Metadata.Catalog.Monitoring;
 using NuGet.Services.Metadata.Catalog.Persistence;
+using NuGet.Services.Storage;
+
+using CatalogIStorageFactory = NuGet.Services.Metadata.Catalog.Persistence.IStorageFactory;
```
Comment (shishirx34) [inline src/Ng/CommandHelpers.cs]: `ICatalogStorageFactory` maybe?

**Comment by scottbommarito** [comment]
@shishirx34 it's done

File: src/Ng/CommandHelpers.cs [hunk 5]
```diff
@@ -363,5 +374,52 @@ public static Func<HttpMessageHandler> GetHttpMessageHandlerFactory(bool verbose
 
             return handlerFunc;
         }
+
+        public static IEnumerable<EndpointFactory.Input> GetEndpointFactoryInputs(IDictionary<string, string> arguments)
+        {
+            return arguments.GetOrThrow<string>(Arguments.EndpointsToTest).Split(';').Select(e => {
```
Comment (shishirx34) [inline src/Ng/CommandHelpers.cs]: Please add a comment on what the endpoints are supposed to look like, some examples.

File: src/Ng/CommandHelpers.cs [hunk 6]
```diff
@@ -363,5 +374,52 @@ public static Func<HttpMessageHandler> GetHttpMessageHandlerFactory(bool verbose
...
+                var storageKeyValue = arguments.GetOrThrow<string>(Arguments.StorageKeyValue);
+                var storageQueueName = arguments.GetOrDefault<string>(Arguments.StorageQueueName);
+
+                var credentials = new StorageCredentials(storageAccountName, storageKeyValue);
+                var account = new CloudStorageAccount(credentials, true);
+                return new AzureStorageQueue<T>(account, storageQueueName,
+                    new JsonStorageQueueMessageSerializer<T>(JsonSerializerUtility.SerializerSettings));
+            }
+
+            throw new ArgumentException($"Unrecognized storageType \"{storageType}\"");
```
Comment (shishirx34) [inline src/Ng/CommandHelpers.cs]: nit: simply couple this exception with the `NotImplementedException` above and throw storageType is not a `AzureStorageType`.

File: src/Ng/Jobs/Monitoring2MonitoringJob.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        {
+        }
+
+        protected override void Init(IDictionary<string, string> arguments, CancellationToken cancellationToken)
+        {
+            var verbose = arguments.GetOrDefault(Arguments.Verbose, false);
+
+            if (arguments.GetOrThrow<string>(Arguments.StorageType) != Arguments.AzureStorageType)
+            {
+                throw new ArgumentException("File storage is not supported!");
```
Comment (shishirx34) [inline src/Ng/Jobs/Monitoring2MonitoringJob.cs]: nit: `Only azure storage is supported.` is more appropriate.

File: src/Ng/Jobs/MonitoringProcessorJob.cs [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+                await RunPackageValidator(queuedContext, token);
+
+                await _queue.Remove(queueMessage, token);
+            }
+            catch (Exception e)
+            {
+                await SaveFailedPackageMonitoringStatus(queuedContext, e, token);
+
+                await _queue.Remove(queueMessage, token);
+            }
```
Comment (shishirx34) [inline src/Ng/Jobs/MonitoringProcessorJob.cs]: move `await _queue.Remove(queueMessage, token);` outside `try..catch`

File: src/Ng/Jobs/MonitoringProcessorJob.cs [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+            var catalogPage = await _client.GetJObjectAsync(catalogPageUri, token);
+            return new CatalogIndexEntry[]
+            {
+                                new CatalogIndexEntry(
+                                    catalogPageUri,
+                                    Schema.DataTypes.PackageDetails.ToString(),
+                                    catalogPage["catalog:commitId"].ToString(),
+                                    DateTime.Parse(catalogPage["catalog:commitTimeStamp"].ToString()),
+                                    id,
+                                    version)
```
Comment (shishirx34) [inline src/Ng/Jobs/MonitoringProcessorJob.cs]: nit: indentation

File: src/Ng/CommandHelpers.cs (hunk 5 above)
Comment (scottbommarito) [reply src/Ng/CommandHelpers.cs]: There are some in `Arguments.EndpointsToTest`.

File: src/Ng/Jobs/MonitoringProcessorJob.cs (hunk 8 above)
Comment (scottbommarito) [reply src/Ng/Jobs/MonitoringProcessorJob.cs]: got it

File: src/Ng/Jobs/MonitoringProcessorJob.cs (hunk 9 above)
Comment (scottbommarito) [reply src/Ng/Jobs/MonitoringProcessorJob.cs]: fixed

File: src/Ng/CommandHelpers.cs (hunk 6 above)
Comment (scottbommarito) [reply src/Ng/CommandHelpers.cs]: good catch

**Comment by scottbommarito** [comment]
going to wait on merging this for a signed queue library to be pushed

**Pull request merged**